### PR TITLE
Add REST API Architecture doc

### PR DESCRIPTION
**Pull request by delventhalz**
Includes tweaks to the REST API and spec:
- POSTs no longer return `200` statuses
- `openapi.yaml` now documents only currently implemented aspects of the API

### CODE REVIEW

File: docs/source/architecture/rest_api.rst [hunk 1]
```diff
@@ -0,0 +1,378 @@
+********
+REST API
+********
+
+*Hyperledger Sawtooth* provides a pragmatic RESTish API for clients to interact
+with a validator using common HTTP/JSON standards. An entirely separate process,
```
Comment (peterschwarz) [inline docs/source/architecture/rest_api.rst]: Suggest changing the sentence starting with "An entirely separate process..." to "It executed in an entirely separate process, which, once running, allows transaction to be submitted...etc"

File: docs/source/architecture/rest_api.rst [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+once running it allows transactions to be submitted and blocks to be read with a
+common language-neutral interface. As the validator is redesigned and improved
+the REST API will grow with it, providing a consistent interface that meets the
+needs of app developers into the future.
+
+With that focus on app developers, the REST API treats the validator mostly as a
+black box, submitting transactions and fetching the results. It is not a tool
+for *all* validator communication. It would not be used by Transaction
```
Comment (peterschwarz) [inline docs/source/architecture/rest_api.rst]: would -> is

File: docs/source/architecture/rest_api.rst [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+       ``"link"`` to check the *status* of the submitted resources is included.
+       If told to wait for commit, but timed out, the status at the moment of
+       timing out is also included in the ``"data"`` property.
+   * - 400
+     - Bad Request
+     - Something in the client's request was malformed, and the request could
+       not be completed.
+   * - 404
+     - Not Found
+     - The request was well formed, but there is no resource with the identifier
```
Comment (peterschwarz) [inline docs/source/architecture/rest_api.rst]: well-formed

File: docs/source/architecture/rest_api.rst [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+=========
+
+The endpoints include RESTful references to resources stored in the Sawtooth
+ledger that clients might be interested in, like blocks and transactions, as
+well as RESTish metadata, like batch status, and validator statistics.
+
+
+Resource Endpoints
+------------------
+In order to fetch resources stored on chain, or in the validator's state,
```
Comment (peterschwarz) [inline docs/source/architecture/rest_api.rst]: drop the comma before "or"

File: docs/source/architecture/rest_api.rst [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+   * **/transactions** - the transactions stored on the blockchain, referenced
+     by id
+   * **/state** - the ledger state, stored on the merkel trie, referenced by
+     leaf addresses
+
+
+Submission Endpoints
+--------------------
+In order to submit transactions to a Sawtooth validator, they *must* be wrapped
+in a batch. For that reason, submissions are sent to the ``/batches`` endpoint,
```
Comment (peterschwarz) [inline docs/source/architecture/rest_api.rst]: drop comma at the end of this line

File: docs/source/architecture/rest_api.rst [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+        {
+          "data": {
+            "8980...029c": "COMMITTED",
+            "540a6...484a": "INVALID",
+            "65cd...47dd": "PENDING"
+          }
+        }
+
+
+Stats Endpoints
```
Comment (peterschwarz) [inline docs/source/architecture/rest_api.rst]: Sounds like this section should be dropped, as it's not implemented.

**Comment by delventhalz** [comment]
Updated architecture document to remove all planned features which are not implemented (i.e. stats endpoints and some query parameters).

File: docs/source/architecture/rest_api.rst [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+*Hyperledger Sawtooth* provides a pragmatic RESTish API for clients to interact
+with a validator using common HTTP/JSON standards. It is an entirely separate
+process, which once running allows transactions to be submitted and blocks to
+be read with a common language-neutral interface. As the validator is redesigned
+and improved the REST API will grow with it, providing a consistent interface
+that meets the needs of app developers into the future.
+
+With that focus on app developers, the REST API treats the validator mostly as a
+black box, submitting transactions and fetching the results. It is not a tool
+for *all* validator communication. It would not be used by Transaction
```
Comment (peterschwarz) [inline docs/source/architecture/rest_api.rst]: "would not be" -> "is not"

File: docs/source/architecture/rest_api.rst [hunk 8]
```diff
@@ -0,0 +1,293 @@
+********
+REST API
+********
+
+*Hyperledger Sawtooth* provides a pragmatic RESTish API for clients to interact
+with a validator using common HTTP/JSON standards. It is an entirely separate
+process, which once running allows transactions to be submitted and blocks to
```
Comment (peterschwarz) [inline docs/source/architecture/rest_api.rst]: comma after "running"

File: docs/source/architecture/rest_api.rst (hunk 7 above)
Comment (delventhalz) [inline docs/source/architecture/rest_api.rst]: The last example (ledger sync), is not currently implemented in anything.

Comment (peterschwarz) [inline docs/source/architecture/rest_api.rst]: Fair point - we should probably remove the ledger sync example, as we replaced it with state export.

File: docs/source/architecture/rest_api.rst [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+Processors to communicate with a validator, or by one validator to talk to other
+validators. For these and similar use cases, there is a more efficient and
+robust, if somewhat more complicated, ZMQ/Protobuf interface.
+
+
+Open API Specification
+======================
+
+The REST API is comprehensively documented using the
+`OpenAPI specification <http://swagger.io/specification/>`_ (fka Swagger),
```
Comment (ojalatodd) [inline docs/source/architecture/rest_api.rst]: typo? What does "fka" mean?

Comment (delventhalz) [inline docs/source/architecture/rest_api.rst]: formally known as

Comment (delventhalz) [inline docs/source/architecture/rest_api.rst]: http://swagger.io/specification/

**Comment by delventhalz** [comment]
Fixed the typo.

**Pull request merged**